### PR TITLE
Use Rye instead of Poetry

### PR DESCRIPTION
**Pull request by Kludex**

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -16,9 +16,7 @@ jobs:
       - uses: actions/setup-python@v5
         with:
           python-version: '3.8'
-      - run: |
-            pip install poetry
-            poetry install -E test
-      - uses: pre-commit/action@v3.0.1
-        with:
-          extra_args: --all-files --verbose
+      - uses: eifinger/setup-rye@v2
```
Comment (samuelcolvin) [inline .github/workflows/main.yml:19]: Surely simplest to just run pre-commit in CI like other projects.

File: pyproject.toml [hunk 2]
```diff
@@ -118,21 +85,23 @@ test = [
...
 ]
 
-[tool.poetry.scripts]
-logfire = "logfire.cli:main"
+[tool.rye.scripts]
+type-check = "pyright"
 
-[tool.poetry.plugins."pydantic"]
-logfire-plugin = "logfire.integrations.pydantic:plugin"
+[build-system]
```
Comment (samuelcolvin) [inline pyproject.toml]: I think this should go right at the top of the file.

File: pyproject.toml [hunk 3]
```diff
@@ -1,89 +1,56 @@
...
-starlette = ["opentelemetry-instrumentation-starlette"]
-sqlalchemy = ["opentelemetry-instrumentation-sqlalchemy"]
-psycopg2 = ["opentelemetry-instrumentation-psycopg2"]
-pymongo = ["opentelemetry-instrumentation-pymongo"]
-redis = ["opentelemetry-instrumentation-redis"]
-requests = ["opentelemetry-instrumentation-requests"]
-test = [
+[tool.rye]
+managed = true
+dev-dependencies = [
```
Comment (samuelcolvin) [inline pyproject.toml:57]: I couldn't work out how to have multiple dev depenency groups, e.g. for docs, but I guess if install is fast enough, it shouldn't matter.

File: .github/workflows/main.yml (hunk 1 above)
Comment (Kludex) [reply .github/workflows/main.yml:19]: It was not simple, I'll add pre-commit here later.

**Pull request merged**